CUDA self-recursion tests

gmarkall opened the pull request:
Self-recursion was mostly already working on the CUDA target due to the recent dispatcher refactoring work. This PR mainly adds tests for it and some fixes to ensure more use cases work correctly. The test cases are directly taken from `numba.tests.test_recursion`, with modifications to adapt them to the CUDA target. Present recursion limitations are:

* Self-recursion only works with the same types - calling the same function recursively with different types does not yet work, because it seems to require insertion of unresolved references, which is not supported by the CUDA target (e.g. `JITCUDACodegen.insert_unresolved_ref()` does not exist).
* Mutual recursion is not supported.

Since I noticed #8036 whilst working on this, there are some additional tests and a small fix related to it here - one of the recursion test cases raises in a device function.

The changes here are:

- Updating the docs to clarify limitations on `assert` and `raise`, and to document the recursion support.
- A fix to `_MinimalCallHelper.get_exception()`, so that it returns the correct number of items. It was supposed to return the location info as well - given that the exception is unknown there's no meaningful info that can be given, but at least returning `None` for it allows the result to be unpacked. With this change, the fact that an unexpected exception occurs can be reported as a `SystemError`, instead of just returning a spurious unpacking error from inside the CUDA target.
- Compilation of device functions registers the dispatcher with type inference, so that recursive calls can be typed.
- The acquisition of `self._compiling_counter` in the CUDA dispatcher is moved into `compile_device()` up from `get_call_template()` - when it was in its former location, some recursive calls were not detected correctly because the compile counter was 0 if `compile_device()` was directly entered rather than coming through `get_call_template()`.
- Adding the `opt` kwarg to the simulator's `jit` decorator - this is needed because some tests pass `opt=False` to the decorator.
- Add a new decorator, `xfail_unless_cudasim` - for example, I would like to mark some of the recursion tests that have issues due to #8036 as xfails, but they succeed on the simulator because it executes the recursion successfully, or it successfully propagates an exception from a device function.
- Addition of self-recursion tests for the CUDA target as detailed above
- Addition of an xfailing test for raising in a device function.

gmarkall commented:
gpuci run tests

gmarkall commented:
gpuci run tests

stuartarchibald commented on numba/cuda/tests/cudapy/test_recursion.py:
```diff
@@ -0,0 +1,123 @@
+class TestSelfRecursion(CUDATestCase):
+
+    def setUp(self):
+        # Avoid importing this module at toplevel, as it triggers compilation
+        # and can therefore fail
```
```suggestion
        # Avoid importing this module at the top level, as it triggers
        # compilation and can therefore fail
```

gmarkall replied:
Applied.

stuartarchibald reviewed (commented):
Thanks for the patch, great to see this working with so few changes! There's a few minor things to resolve but in general looks good. Thanks again!

gmarkall commented on numba/cuda/dispatcher.py:
```diff
@@ -662,22 +661,21 @@ def get_call_template(self, args, kws):
-            # because we may be able to force a cast on the caller side.
-            if self._can_compile:
-                self.compile_device(tuple(args))
+        # Ensure an exactly-matching overload is available if we can
```
The differences seem to be:

- The base class also calls `pysig, args = self._compiler.fold_argument_types(args, kws)`
- This implementation needs to call `compile_device()` instead of `compile()`
- Because this implementation didn't call `fold_argument_types()`, it needs to use `utils.pysignature()` to get `pysig`.

I think it's a bit too different to consider uniting them at this stage.

gmarkall commented on numba/cuda/dispatcher.py:
```diff
@@ -688,27 +686,28 @@ def compile_device(self, args):
-            cres.target_context.insert_user_function(cres.entry_point,
-                                                     cres.fndesc,
-                                                     [cres.library])
+            with self._compiling_counter:
```
It's not clear to me that the logic isn't backwards in the base class - shouldn't the base class not be incrementing the compiling counter if it's only going to return an existing overload?

(Or to put it another way - isn't the base class the one with backwards logic, and it too should check for an existing overload prior to incrementing the compiling counter?)

stuartarchibald replied:
@sklam Think you wrote this originally, do you recall why it's this way around? If not, I'll go take a look at what's going on in the base/CPU target, it could be an oversight/bug.

gmarkall replied:
To add a little more detail - I think it doesn't technically matter where the check for / return of the existing overload is, because the compiling counter is only used to detect recursion during another compilation that could get started during the first one. Since we won't start compiling if an overload does exist, it doesn't matter whether the compiling counter has been incremented or not (and if it was incremented, it will be decremented again as soon as the existing overload is returned).

Note, I do prefer my variant of the logic where the counter isn't incremented at all if there's an existing overload, because it seems to conflate fewer things at a time.

sklam replied:
issue at https://github.com/numba/numba/issues/8162

gmarkall commented:
gpuci run tests

gmarkall commented on numba/core/callconv.py:
```diff
@@ -315,7 +315,10 @@ def get_exception(self, exc_id):
             return self.exceptions[exc_id]
         except KeyError:
             msg = "unknown error %d in native function" % exc_id
-            return SystemError, (msg,)
```
No problem - docstrings updated for the `_MinimalCallHelper` class.

gmarkall commented on docs/source/cuda/cudapysupported.rst:
```diff
@@ -54,6 +56,44 @@ This is due to a general limitation in CUDA printing, as outlined in the
+---------
+
+Self-recursive device functions are supported, with the constraint that
+recursive calls must all have similar argument types. For example, the following
```
I've adjusted the wording - I tried to avoid the word "signature" as I thought that might not have a clear meaning to the average reader, but hopefully it conveys the required meaning now.

gmarkall commented on numba/cuda/tests/cudapy/recursion_usecases.py:
```diff
@@ -0,0 +1,93 @@
+from numba import cuda
```
Added.

gmarkall commented on numba/cuda/tests/cudapy/test_recursion.py:
```diff
@@ -0,0 +1,123 @@
+
+            expected = pfunc(arg)
+            if expected is None:
+                expected = 999
```
Good point, I did feel that the way I'd written this was weird, but didn't spot the simplification. Thanks! Test now updated.

gmarkall commented on docs/source/cuda/cudapysupported.rst:
```diff
@@ -54,6 +56,44 @@ This is due to a general limitation in CUDA printing, as outlined in the
+
+Mutual recursion between functions (e.g. where a function ``func1()`` calls
+``func2()`` which again calls ``func1()``) is unsupported.
+
```
Note added.

gmarkall commented on docs/source/cuda/cudapysupported.rst:
```diff
@@ -54,6 +56,44 @@ This is due to a general limitation in CUDA printing, as outlined in the
+           return n
+       return fib(n - 1) + fib(n - 2)
+
+(the ``fib`` function always has an ``int64`` argument), wherease the following
```
Suggestion applied.

gmarkall commented on numba/cuda/simulator/api.py:
```diff
@@ -77,7 +77,7 @@ def elapsed_time(self, event):
 def jit(func_or_sig=None, device=False, debug=False, argtypes=None,
         inline=False, restype=None, fastmath=False, link=None,
-        boundscheck=None,
+        boundscheck=None, opt=None
```
It should be consistent - I've updated the simulator.

gmarkall commented:
@stuartarchibald Many thanks for the review - I believe I've addressed all comments. The CI fails seem to be some sort of conda / dependency issue unrelated to this PR, so I think it's ready for another look.

stuartarchibald commented:
> @stuartarchibald Many thanks for the review - I believe I've addressed all comments. The CI fails seem to be some sort of conda / dependency issue unrelated to this PR, so I think it's ready for another look.

@gmarkall Many thanks. Any chance you could resolve the merge conflict please? Pushing that patch should also re-run CI against `main` which now has the issue with the conda package dependencies fixed (well, worked around!). Thanks again!

gmarkall commented:
@stuartarchibald thanks for looking, conflicts now resolved.

gpuci run tests

sklam reviewed (approved):
Thanks for the patch!

stuartarchibald commented:
Buildfarm ID: `numba_smoketest_cuda_yaml_136`.
